### PR TITLE
enable Symfony 3.0

### PR DESCRIPTION
**Pull request by lsmith77**
fix for https://github.com/FriendsOfSymfony/FOSRestBundle/issues/1106

### CODE REVIEW

**Comment by lsmith77** [comment]
```
PHP Fatal error:  Call to undefined method Mock_ValidatorInterface_5cfaa869::validate() in /home/travis/build/FriendsOfSymfony/FOSRestBundle/Request/AbstractRequestBodyParamConverter.php on line 159
```

**Comment by xabbuh** [comment]
The `testLegacyValidatorParameters()` in `AbstractRequestBodyParamConverterTest` uses the deprecated validator API that will be removed with Symfony 3.0. This test can probably simply be skipped when the legacy API is not available (i.e. `Symfony\Component\Validator\ValidatorInterface` does not exist).

**Comment by lsmith77** [comment]
```
There were 2 errors:
1) FOS\RestBundle\Tests\View\ViewHandlerTest::testCreateResponseWithFormErrorsAndSerializationGroups with data set "json" ('json')
Symfony\Component\Form\Exception\InvalidArgumentException: Could not load type "text"
/home/travis/build/FriendsOfSymfony/FOSRestBundle/vendor/symfony/symfony/src/Symfony/Component/Form/FormRegistry.php:91
/home/travis/build/FriendsOfSymfony/FOSRestBundle/vendor/symfony/symfony/src/Symfony/Component/Form/FormFactory.php:84
/home/travis/build/FriendsOfSymfony/FOSRestBundle/vendor/symfony/symfony/src/Symfony/Component/Form/FormBuilder.php:106
/home/travis/build/FriendsOfSymfony/FOSRestBundle/vendor/symfony/symfony/src/Symfony/Component/Form/FormBuilder.php:267
/home/travis/build/FriendsOfSymfony/FOSRestBundle/vendor/symfony/symfony/src/Symfony/Component/Form/FormBuilder.php:215
/home/travis/build/FriendsOfSymfony/FOSRestBundle/Tests/View/ViewHandlerTest.php:598
2) FOS\RestBundle\Tests\View\ViewHandlerTest::testCreateResponseWithFormErrorsAndSerializationGroups with data set "xml" ('xml')
Symfony\Component\Form\Exception\InvalidArgumentException: Could not load type "text"
/home/travis/build/FriendsOfSymfony/FOSRestBundle/vendor/symfony/symfony/src/Symfony/Component/Form/FormRegistry.php:91
/home/travis/build/FriendsOfSymfony/FOSRestBundle/vendor/symfony/symfony/src/Symfony/Component/Form/FormFactory.php:84
/home/travis/build/FriendsOfSymfony/FOSRestBundle/vendor/symfony/symfony/src/Symfony/Component/Form/FormBuilder.php:106
/home/travis/build/FriendsOfSymfony/FOSRestBundle/vendor/symfony/symfony/src/Symfony/Component/Form/FormBuilder.php:267
/home/travis/build/FriendsOfSymfony/FOSRestBundle/vendor/symfony/symfony/src/Symfony/Component/Form/FormBuilder.php:215
/home/travis/build/FriendsOfSymfony/FOSRestBundle/Tests/View/ViewHandlerTest.php:598
--
There were 11 failures:
1) FOS\RestBundle\Tests\Functional\SerializerErrorTest::testSerializeExceptionJson with data set #0 ('Serializer', '{"code":500,"message":"Someth...:null}')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'{"code":500,"message":"Something bad happened.","errors":null}'
+'An Exception was thrown while handling: Something bad happened.'
/home/travis/build/FriendsOfSymfony/FOSRestBundle/Tests/Functional/SerializerErrorTest.php:22
2) FOS\RestBundle\Tests\Functional\SerializerErrorTest::testSerializeExceptionJson with data set #1 ('JMSSerializer', '{"code":500,"message":"Someth...ned."}')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'{"code":500,"message":"Something bad happened."}'
+'An Exception was thrown while handling: Something bad happened.'
/home/travis/build/FriendsOfSymfony/FOSRestBundle/Tests/Functional/SerializerErrorTest.php:22
3) FOS\RestBundle\Tests\Functional\SerializerErrorTest::testSerializeExceptionXml with data set #0 ('Serializer', '<?xml version="1.0"?>\n<respon...onse>\n')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'<?xml version="1.0"?>
-<response><code>500</code><message>Something bad happened.</message><errors/></response>
-'
+'An Exception was thrown while handling: Something bad happened.'
/home/travis/build/FriendsOfSymfony/FOSRestBundle/Tests/Functional/SerializerErrorTest.php:43
4) FOS\RestBundle\Tests\Functional\SerializerErrorTest::testSerializeExceptionXml with data set #1 ('JMSSerializer', '<?xml version="1.0" encoding=...sult>\n')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'<?xml version="1.0" encoding="UTF-8"?>
-<result xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
-  <code>500</code>
-  <message><![CDATA[Something bad happened.]]></message>
-  <errors xsi:nil="true"/>
-</result>
-'
+'An Exception was thrown while handling: Something bad happened.'
/home/travis/build/FriendsOfSymfony/FOSRestBundle/Tests/Functional/SerializerErrorTest.php:43
5) FOS\RestBundle\Tests\Functional\SerializerErrorTest::testSerializeInvalidFormJson with data set #0 ('Serializer', '{"code":400,"message":"Valida..."]}}}}')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'{"code":400,"message":"Validation Failed","errors":{"children":{"name":{"errors":["This value should not be blank."]}}}}'
+'An Exception was thrown while handling: Could not load type "text"'
/home/travis/build/FriendsOfSymfony/FOSRestBundle/Tests/Functional/SerializerErrorTest.php:78
6) FOS\RestBundle\Tests\Functional\SerializerErrorTest::testSerializeInvalidFormJson with data set #1 ('JMSSerializer', '{"code":400,"message":"Valida..."]}}}}')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'{"code":400,"message":"Validation Failed","errors":{"children":{"name":{"errors":["This value should not be blank."]}}}}'
+'An Exception was thrown while handling: Could not load type "text"'
/home/travis/build/FriendsOfSymfony/FOSRestBundle/Tests/Functional/SerializerErrorTest.php:78
7) FOS\RestBundle\Tests\Functional\SerializerErrorTest::testSerializeInvalidFormXml with data set #0 ('Serializer', '<?xml version="1.0"?>\n<respon...onse>\n')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'<?xml version="1.0"?>
-<response><code>400</code><message>Validation Failed</message><errors><children><name><errors>This value should not be blank.</errors></name></children></errors></response>
-'
+'An Exception was thrown while handling: Could not load type "text"'
/home/travis/build/FriendsOfSymfony/FOSRestBundle/Tests/Functional/SerializerErrorTest.php:97
8) FOS\RestBundle\Tests\Functional\SerializerErrorTest::testSerializeInvalidFormXml with data set #1 ('JMSSerializer', '<?xml version="1.0" encoding=...sult>\n')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'<?xml version="1.0" encoding="UTF-8"?>
-<result>
-  <code>400</code>
-  <message><![CDATA[Validation Failed]]></message>
-  <errors>
-    <form name="form">
-      <errors/>
-      <form name="name">
-        <errors>
-          <entry><![CDATA[This value should not be blank.]]></entry>
-        </errors>
-      </form>
-    </form>
-  </errors>
-</result>
-'
+'An Exception was thrown while handling: Could not load type "text"'
/home/travis/build/FriendsOfSymfony/FOSRestBundle/Tests/Functional/SerializerErrorTest.php:97
9) FOS\RestBundle\Tests\Request\RequestBodyParamConverterTest::testExecutionInterceptsSymfonySerializerException
Failed asserting that exception of type "Symfony\Component\Serializer\Exception\InvalidArgumentException" matches expected exception "Symfony\Component\HttpKernel\Exception\BadRequestHttpException". Message was: "" at
#0 [internal function]: FOS\RestBundle\Tests\Request\AbstractRequestBodyParamConverterTest->testExecutionInterceptsSymfonySerializerException()
#1 phar:///home/travis/.phpenv/versions/5.5.21/bin/phpunit/phpunit/Framework/TestCase.php(860): ReflectionMethod->invokeArgs(Object(FOS\RestBundle\Tests\Request\RequestBodyParamConverterTest), Array)
#2 phar:///home/travis/.phpenv/versions/5.5.21/bin/phpunit/phpunit/Framework/TestCase.php(737): PHPUnit_Framework_TestCase->runTest()
#3 phar:///home/travis/.phpenv/versions/5.5.21/bin/phpunit/phpunit/Framework/TestResult.php(609): PHPUnit_Framework_TestCase->runBare()
#4 phar:///home/travis/.phpenv/versions/5.5.21/bin/phpunit/phpunit/Framework/TestCase.php(693): PHPUnit_Framework_TestResult->run(Object(FOS\RestBundle\Tests\Request\RequestBodyParamConverterTest))
#5 phar:///home/travis/.phpenv/versions/5.5.21/bin/phpunit/phpunit/Framework/TestSuite.php(716): PHPUnit_Framework_TestCase->run(Object(PHPUnit_Framework_TestResult))
#6 phar:///home/travis/.phpenv/versions/5.5.21/bin/phpunit/phpunit/Framework/TestSuite.php(716): PHPUnit_Framework_TestSuite->run(Object(PHPUnit_Framework_TestResult))
#7 phar:///home/travis/.phpenv/versions/5.5.21/bin/phpunit/phpunit/TextUI/TestRunner.php(398): PHPUnit_Framework_TestSuite->run(Object(PHPUnit_Framework_TestResult))
#8 phar:///home/travis/.phpenv/versions/5.5.21/bin/phpunit/phpunit/TextUI/Command.php(152): PHPUnit_TextUI_TestRunner->doRun(Object(PHPUnit_Framework_TestSuite), Array)
#9 phar:///home/travis/.phpenv/versions/5.5.21/bin/phpunit/phpunit/TextUI/Command.php(104): PHPUnit_TextUI_Command->run(Array, true)
#10 /home/travis/.phpenv/versions/5.5.21/bin/phpunit(722): PHPUnit_TextUI_Command::main()
#11 {main}.
10) FOS\RestBundle\Tests\Routing\Loader\RestRouteLoaderTest::testAnnotatedUsersFixture
requirements failed to match for get_users
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    '_method' => 'GET'
 )
/home/travis/build/FriendsOfSymfony/FOSRestBundle/Tests/Routing/Loader/RestRouteLoaderTest.php:109
11) FOS\RestBundle\Tests\Routing\Loader\RestRouteLoaderTest::testAnnotatedConditionalUsersFixture
requirements failed to match for get_users
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    '_method' => 'GET'
 )
/home/travis/build/FriendsOfSymfony/FOSRestBundle/Tests/Routing/Loader/RestRouteLoaderTest.php:155
                                                      
FAILURES!                                             
Tests: 318, Assertions: 1422, Failures: 11, Errors: 2.
```

**Comment by lsmith77** [comment]
looks like there are some issues with form type loading and exception handling

**Comment by WouterJ** [comment]
Are you aware that form type names are removed in sf 3 in favor of the fqcn of the type class?

**Comment by lsmith77** [comment]
nope. so any way i can make this work in both 2.7 and 3.0?

**Comment by xabbuh** [comment]
You can check for the [getBlockPrefix()](https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Form/AbstractType.php#L82-L89) method being present in the `AbstractType` class.

File: composer.json [hunk 1]
```diff
@@ -32,16 +32,16 @@
...
-        "symfony/browser-kit":            "^2.7",
-        "symfony/dependency-injection":   "^2.7",
-        "symfony/expression-language":    "~2.7",
-        "symfony/css-selector":           "^2.7",
+        "symfony/phpunit-bridge":         "~2.7|^3.0",
+        "symfony/form":                   "^2.7|^3.0",
+        "symfony/validator":              "^2.7|^3.0",
+        "symfony/serializer":             "^2.7|^3.0",
+        "symfony/yaml":                   "^2.7|^3.0",
+        "symfony/security":               "^2.|^3.07",
```
Comment (xabbuh) [inline composer.json]: This looks wrong (see #1145).

**Pull request merged**